tests/p2p/test_peer.py

```diff
@@ -52,6 +53,30 @@ def test_sub_protocol_selection():
         peer.select_sub_protocol([('unknown', 1)])
 
 
+@pytest.mark.asyncio
+async def test_peer_pool_iter(request, event_loop):
+    peer1, _ = await get_directly_linked_peers(request, event_loop)
+    peer2, _ = await get_directly_linked_peers(request, event_loop)
+    peer3, _ = await get_directly_linked_peers(request, event_loop)
+    pool = MockPeerPoolWithConnectedPeers([peer1, peer2, peer3])
+    peers = list([peer async for peer in pool])
+
+    assert len(peers) == 3
+    assert peer1 in peers
+    assert peer2 in peers
+    assert peer3 in peers
+
+    peers = []
+    asyncio.ensure_future(peer2.disconnect(DisconnectReason.disconnect_requested))
+    async for peer in pool:
+        peers.append(peer)
+
+    assert len(peers) == 2
+    assert peer1 in peers
+    assert peer2 not in peers
```

Comment:
Should there be an assertion that `peer2` is still found in the connected nodes?  Otherwise we aren't testing what we think we're testing if the disconnect has already been processed.

Reply:
It won't be in `connected_nodes`, but that's not really what we want to ensure with this test. The point of this change is to force callsites to iterate asynchronously over the peers in the pool so that on every iteration we can process disconnections, which in turn allow us to skip stale items from the _snapshot_ we get when the iteration starts -- and this test is ensuring we're skipping those stale items from the snapshot.